HSEARCH-3319 Type-safe field references / HSEARCH-5300 Annotation processor for the Search static metamodel (#4156)
hibernate/hibernate-search: documentation/src/main/asciidoc/public/reference/_static-metamodel-processor.adoc

                        <artifactId>hibernate-search-metamodel-processor</artifactId> <2>
                    </path>
                    <path>
                        <groupId>org.hibernate.search</groupId>
                        <artifactId>hibernate-search-backend-lucene</artifactId> <3>
                    </path>
                </annotationProcessorPaths>
            </configuration>
        </execution>
    </executions>
</plugin>
----

<1> Provide the fully qualified class name of the annotation processor that generates the metamodel.
<2> Add the `org.hibernate.search:hibernate-search-metamodel-processor` dependency as an annotation processor path for the Maven compiler plugin to be able to actually find the processor.
<3> Add the backend dependency, in this example, the <<backend-lucene,Lucene backend>>, as an annotation processor path.
It is important to include the same backend that the application is using to make sure that the generated metamodel classes reflect all the backend specifics.
For example, backends might have different defaults, resulting in a different set of field traits depending on the backend.

> NOTE: The version of both annotation processor and backend dependencies are skipped in the definition of the annotation paths, as in case the Hibernate Search BOM is imported via dependency management, the compiler plugin will use the managed versions.

Reviewer: ```suggestion
NOTE: The version of both annotation processor and backend dependencies can be ommitted in the definition of the annotation paths, because they are defined in the Hibernate Search BOM, which we recommend you import via dependency management.
```

This way the generated metamodel classes will be based on the same backend that the application uses.
====

[[static-metamodel-processor-configuration]]
== Configuration

The annotation processor options are passed as the compiler arguments with the `-A` key:

====
[source,XML,subs="+attributes"]
----
<plugin>
    <artifactId>maven-compiler-plugin</artifactId>
    <executions>
        <execution>
            <id>default-compile</id>
            <configuration>
                <compilerArgs>
                    <arg>-Aorg.hibernate.search.metamodel.processor.generated_annotation.timestamp=false</arg> <1>
                    <arg><!-- ... --></arg> <2>
                </compilerArgs>
                <!-- ... --> <3>
            </configuration>
        </execution>
    </executions>
</plugin>
----

<1> Pass the annotation processor parameters using the `-A` key.
<2> Pass any other compiler arguments required by the build.
<3> Further compiler plugin configuration.
====

The following annotation processor configuration properties are available:

> [[static-metamodel-processor-configuration-generated_annotation-add]]`org.hibernate.search.metamodel.processor.generated_annotation.add`::

Reviewer: Same here, you might want to drop the `.metamodel` part?

Description:::
Whether to add the `@Generated` annotation to the generated static metamodel classes.
Default value:::
`true`

[[static-metamodel-processor-configuration-generated_annotation-timestamp]]`org.hibernate.search.metamodel.processor.generated_annotation.timestamp`::
Description:::
Defines whether the `@Generated` annotation includes the `date` attribute.
Having the date attribute will result in non-reproducible builds, as the timestamp will be different for each compilation.
Hence, it is disabled by default.
Default value:::
`false`

[[static-metamodel-processor-configuration-add_generated_annotation]]`org.hibernate.search.metamodel.processor.backend.version`::
Description:::

> Explicitly define the backend version. By default, the processor will use the latest compatible version of the backend.

Reviewer: Do you mean the Elasticsearch version?

If so, you should probably be more specific, and maybe link to the relevant section of the docs.

Author: It's passed to both backends, while it probably makes close to 0 difference for the Lucene one ... it still will pass it over, who knows if we will have some dependency on the index version 😃.

This option can be used if the static metamodel is required for an older backend version.
Default value:::
`<latest>`

[[static-metamodel-processor-limitations]]
== Current annotation processor limitations

> While the annotation processor is in its early stages of development, it has a few limitations:

Reviewer: ```suggestion
While the annotation processor is very much functional and we encourage you to try it, it is still in the early stages of development, and thus has a few limitations:
```

* Any use of <<binding-valuebridge-valuebinder,custom binders>> will be ignored and should produce a compiler warning.
This means that if the search entities rely on custom binders, fields that those binders produce will be missing from the generated metamodel.

> * <<mapper-orm-custom-annotations,Custom mapping annotations>> are not supported.

Reviewer: ```suggestion
* <<mapper-orm-custom-annotations,Custom mapping annotations>> are ignored without warning.
```

* <<mapping-programmatic,Programmatic mapping>> is also unsupported by the annotation processor.

> As the annotation processor cannot easily read programmatic mapping definitions, this limitation is here to stay.

Reviewer: ```suggestion
As the annotation processor cannot possibly execute programmatic mapping defined in the code that is being compiled, this limitation is here to stay.
```
